### ISSUE TITLE
PyGraphistry plotter integration

### ISSUE THREAD

**Issue by lmeyerov** [igraph/python-igraph#536]
**What is the feature or improvement you would like to see?**

igraph shortcuts & docs for use with pygraphistry for GPU plotting & GNN/UMAP features, basically mirroring what we've been doing for pygraphistry->igraph

I can try assisting here -- am curious what'd be useful? Any guidance etc. would be appreciated!

**Use cases for the feature**

* Scalable, highly-interactive, & interactive plotting is useful

* Streamlined integration & docs to solve data wrangling ergonomics would benefit both advanced and novice users

* Maintaining bindings on both sides would also help catch regressions faster

**Sample interfaces**

Similar to the networkx/graphtool/... bindings, I can imagine:

```python
pip install python-igraph graphistry
ig.to_graphistry().plot()  #and other pygraphistry methods -- color encodings, embeddings, ... 
```

I'm less clear on how the multiple plotting engines are supported now, I'm guessing something like:

```python
pip install python-igraph[plotting]
ig.plot(g, layout=layout, engine='graphistry')

# how does .plot() know which engine? is there a `ig.set_plot_engine()`?
```

**References**

Recent igraph integration on the pygraphistry side:
* Ex: Using igraph layouts with graphistry visualizations: https://github.com/graphistry/pygraphistry/pull/346
* Ex: Using igraph analytics with graphistry visualizations:  https://github.com/graphistry/pygraphistry/blob/master/demos/more_examples/simple/MarvelTutorial.ipynb

cc @iosonofabio  @silkspace

**Comment by ntamas** [comment]
This would be a very welcome addition indeed. In the past we've had a custom `GraphDrawer` subclass for [Ubigraph](https://github.com/0x4lan/ubigraph_server), but the the development of the Ubigraph server stopped around 2008 and the graph drawer on our side went defunct. It would be great if we could revive the functionality with PyGraphistry.

> I'm less clear on how the multiple plotting engines are supported now

The key to adding a new plotting backend is to derive your own graph drawer subclass from `igraph.drawing.baseclasses.AbstractGraphDrawer`. Currently there are two options to select the backend: either you instantiate a graph drawer class on your own and use its `draw()` method directly, or you call `plot(graph, target=...)` and the type of the target decides the plotting backend. Right now the Matplotlib backend is selected if the target is a Matplotlib axes object, otherwise we use the Cairo plotting backend.

The `develop` branch became a bit more sophisticated with the backend selection:

* this branch also supports `plotly` and will use the Plotly backend if the plot target is a `plotly.graph_objects.Figure` object
* if the target is not a Matplotlib axes object, not a Plotly figure and not a Cairo surface, then the system picks the backend from the `plotting.backend` key of igraph's global configuration object; this is meant for interactive environments.

Unfortunately the logic for backend selection is now buried in the `plot()` function; I think it should be refactored out to a class method of the `AbstractGraphDrawer` class that accepts an arbitrary object and returns whether the object can be handled by the backend.

Anyhow, I don't know much about Graphistry yet (need to do the research on my own), but in order to start with Graphistry integration, I think the best would be to pick a Graphistry object similar to `matplotlib.Axes` or `plotly.graph_objects.Figure`, create a `GraphistryGraphDrawer` class, and update the `plot()` method to use the Graphistry graph drawer backend if the target object is the Graphistry-specific object that you've picked. It is best to work in the `develop` branch because the plotting parts have undergone heavy refactoring in the last few months and the `master` branch is unlikely to get more updates.

Summon @iosonofabio for more input as he was doing most of the refactoring and he probably knows the best approach better than me by now.

**Comment by iosonofabio** [comment]
Summoning... success!

Both points (export to external library and plotting) were PRs of mine in `python-igraph`, happy to help.

- **Export**: this seems easy, we can basically do:
```python
def to_graphistry(self, ...whatever optional args...):
    import pygraphistry
    return pygraphistry.from_igraph(self, ...args...)
```

which is trivial **and** useful. Let's just make a PR for this.

- **Plotting**: the current backends (Cairo, matplotlb, plotly) do not have native graph plotting. If graphistry does, we should not reinvent the wheel and just write a thin shim to convert plotting arguments and graph attributes into graphistry instructions and then use your own plotting infra:
```python
def plot(obj, target=None, ...):
   ...
   elif isinstance(target, graphistry.FigureCanvas) or (backend == 'graphistry'):
       return obj.to_graphistry(... options ...).plot(target)
```

I agree with @ntamas, of course, that the logic of backend choice should perhaps not be buried in the plotting function itself. But I think that's a separate concern of ours that we can sort out independently, to keep things modular.

In conclusion, here's my 2 cents:
1. start a PR for the export (see above)
2. once that's merged, start a PR for the plotting, editing only the plot function in `src/igraph/drawing/__init__.py` as above, and I'll help you take it from there
3. once the second PR is merged, I'll buy a broom and sweep out the plotting backend/engine logic from that plot function into a more appropriate corner of the library

**Comment by lmeyerov** [comment]
Makes sense!

I'll start a new PR1 for `to_graphistry()`, including tests and docs, and then we can revisit `plot()` . We're getting a big release out on our side so probably not till the weekend.

Quick q: should it fork main or some specific edge branch, as it sounds like some things are moving?

**Comment by lmeyerov** [comment]
(fwiw, still of interest there, just some interim priorities)
